### PR TITLE
Make ZIL operations on zvols use _by_dnode routines

### PR DESCRIPTION
**Pull request by ryao**
### Description

This continues what was started in
0eef1bde31d67091d3deed23fe2394f5a8bf2276 by fully converting zvols to
avoid unnecessary dnode_hold() calls. This saves a small amount of CPU
time and slightly improves latencies onsynchronous operations on zvols.

### Motivation and Context

I am working on zvol performance at work. While working on larger changes, I noticed unnecessary overhead that is easy to eliminate. Eliminating it brings some small latency improvements from reduced overhead.

### How Has This Been Tested?

It has been tested against prophetstor's internal 0.6.5.9 branch using fio:

```
fio --name=writeiops --filename=/dev/zvol/e7c61cba0e8e4cec8692570e7c9280bf/test --direct=1 --rw=randwrite --bs=4k --numjobs=4 --iodepth=32 --direct=1 --iodepth_batch=16 --iodepth_batch_complete=16 --runtime=1200 --ramp_time=5 --norandommap --time_based --ioengine=libaio --group_reporting --random_distribution=pareto:0.9
```

The test hardware is 12x M630-DC SSDs, 256GB of RAM and dual Intel(R) Xeon(R) CPU E5-2630 v4 CPUs (20 cores). I tested on a 500GB zvol. I was careful to wait until the cache was hot enough that we weren't reading metadata before evaluating impact.

The branch is using a variation of #5824 (that predates my first day at my new employer last week), so the numbers are not directly applicable to head, but average latencies dropped by ~15%.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [x] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by ryao** [comment]
@behlendorf I have tested this change on Prophetstor's internal branch based on 0.6.5.9. The internal branch uses the old zvol taskq code (this decision predates me and is something I plan to remedy in the very near future). The hardware is 12x M630-DC SSDs, 256GB of RAM and dual Intel(R) Xeon(R) CPU E5-2630 v4 CPUs (20 cores). I tested using a 500GB zvol by running `fio --name=writeiops --filename=/dev/zvol/e7c61cba0e8e4cec8692570e7c9280bf/test --direct=1 --rw=randwrite --bs=4k --numjobs=4 --iodepth=32 --direct=1 --iodepth_batch=16 --iodepth_batch_complete=16 --runtime=1200 --ramp_time=5 --norandommap --time_based --ioengine=libaio --group_reporting --random_distribution=pareto:0.9`. There are no significant changes to performance (although I suspect latencies might be slightly lower). However, perf profiling indicates less CPU time is being spent.

**Comment by ryao** [comment]
I am working on bigger changes, but this should help #4880.

**Comment by ryao** [comment]
@behlendorf The failure is #6075, which is a pre-existing issue. It is unrelated to the patch here.

**Comment by ryao** [comment]
@sempervictus If you have a synchronous IO intensive workload, this might give you a small boost. You might want to try testing it.

**Comment by sempervictus** [comment]
Oh yeah, going into the weekend stack for sure. Thank you sir :-)

**Comment by behlendorf** [comment]
@ryao if you want these improvements to make 0.7.0 can you address the review feedback.  We're trying to wrap up the release.

**Comment by ryao** [comment]
@behlendorf This is now refreshed.

**Comment by behlendorf** [comment]
@ryao one last rebase to resolve a merge conflict and this should be good to go.

**Comment by ryao** [comment]
@behlendorf It is rebased.

**Comment by behlendorf** [comment]
@ryao thanks, I'll get this merged after the buildbot finishes its work.

**Comment by sempervictus** [comment]
Those test failures dont look too good, crash seems to be coming from:
```
[ 4304.650875] VERIFY(dn->dn_type != DMU_OT_NONE) failed
[ 4304.655793] PANIC at dbuf.c:2303:dbuf_create()
```
which seems related to this :)

**Comment by behlendorf** [comment]
@sempervictus I can see why you thought this might be related since it paniced in `dbuf_create()`, but it's not.  The panic occurred in the zil code for a filesystem and not a zvol, it's more likely this is related to 1b7c1e5ce90ae27d9bb1f6f3616bf079c168005c which was recently merged.  [The link for the log](http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20Mainline%20%28TEST%29/builds/2741/steps/shell_8/logs/log) in the previous comment.

**Comment by sempervictus** [comment]
Yeah... That zil PR should probably be backed out or fixed ASAP - my builders are all OK, and first ztest is usually fine, but zloops are all hanging with no stack trace. Often on fletcher inc test according to log (well, last output is fletcher, next one would be fletcher inc). Something in there is not playing well on Linux in ways more subtle than initial automated testing could detect. It also conflicts in what appear to be key areas with @ryao's zil PR, which as I understand it has been beaten up on Linux pretty heavily. Maybe we switch em out and do the conflict resolution from upstream to zol as opposed to making @ryao change his code to match something that's broken?

**Comment by behlendorf** [comment]
@sempervictus are you able to reproduce those `zloop` failures on the current master branch as of commit 88c30122e5ccb343c5675f4429ee29fc5ba224a0.  If so can you determine as of which commit you started seeing them.

File: module/zfs/dmu.c [hunk 1]
```diff
@@ -965,6 +965,9 @@ int
 dmu_read_by_dnode(dnode_t *dn, uint64_t offset, uint64_t size, void *buf,
     uint32_t flags)
 {
+	if (size == 0)
+		return (0);
```
Comment (behlendorf) [inline module/zfs/dmu.c]: Why was this added?  I don't see why it should be needed here or in the write path.

Comment (ryao) [inline module/zfs/dmu.c]: The previous `dmu_read_uio_dbuf()` function did it. I'll be happy to move it to the write path.

**Comment by ryao** [comment]
@behlendorf The failures seem to have disappeared. Why they appeared when I pushed this initally is something of a mystery, but we are passing now. I have made all of the changes you requested. It should be fine to merge now.

**Comment by behlendorf** [comment]
Thanks @ryao, I'll get this merged.

It turns out the [original failure](http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20Mainline%20%28TEST%29/builds/2741/steps/shell_8/logs/log) reported by @sempervictus wasn't introduced by this PR.  I've seen it one other time against master.  I believe it was either introduced by commit 1b7c1e5 or possibly 38240eb.  As soon as we can determine which commit actually introduced it we're going to want to either revert that change or if it's straight forward fix it.

I also resubmitted the two failed test runs for the updated version of this patch since they hit unrelated issues.  I just wanted to confirm that before merging.
